### PR TITLE
storage: resolve various followups from #15477

### PR DESCRIPTION
**Pull request by guswynn**
See comments on #15477.

That pr had to merge quickly for the release.

### CODE REVIEW

File: src/storage/src/source/antichain.rs [hunk 1]
```diff
@@ -103,8 +104,7 @@ impl OffsetAntichain {
 
     // Frontier apis
 
-    /// Insert a new `MzOffset` frontier value for the `pid`, returning
-    /// the old one if it wasn't there.
+    /// Insert a new `MzOffset` frontier value for
```
Comment (aljoscha) [inline src/storage/src/source/antichain.rs:107]: 🤔

File: src/storage/src/source/antichain.rs [hunk 2]
```diff
@@ -284,11 +286,33 @@ impl PartitionOffset {
...
+    }
+}
+
+// NOTE: Ord is only required as an implementation detail of `MutableAntichain`,
+// but it feels iffy.
+impl Ord for PartitionOffset {
+    fn cmp(&self, other: &Self) -> Ordering {
+        match self.partition.cmp(&other.partition) {
+            Ordering::Equal => self.offset.cmp(&other.offset),
+            other => other,
```
Comment (aljoscha) [inline src/storage/src/source/antichain.rs:315]: nit: the use of `other` is a bit confusing, but ... 🤷

File: src/storage/src/source/source_reader_pipeline.rs [hunk 3]
```diff
@@ -1133,6 +1135,19 @@ where
...
+                    // This should always be true:
+                    // we only advance the `global_source_upper` PAST
+                    // a batch when that batch is fully processed. Offsets are required
+                    // (and in fact, we assert this above) to increase or stay the same
+                    // as we produce batches and a batch upper is always exactly 1
+                    // greater than the maximum offset in that batch. These
+                    // combined mean that the upper of one batch can never be greater
+                    // than the lower of the next.
+                    //
+                    // Our use of `source_upper_updates` ensures the `global_source_upper`
```
Comment (aljoscha) [inline src/storage/src/source/source_reader_pipeline.rs:1147]: It's not necessarily the same order, but we only process _source_upper updates_ once the frontier has advanced, ensuring that we have seen all batches that would have had a lower frontier.

**Review by aljoscha — APPROVED** [review]
Looks good! I had two clarifications, though. Would be premo if you could address these before merging.

File: src/storage/src/source/antichain.rs (hunk 1 above)
Comment (guswynn) [reply src/storage/src/source/antichain.rs:107]: no idea what happened here

**Pull request merged**